File: common-core/src/main/java/io/quartic/geojson/GeoJsonParser.java
package io.quartic.geojson;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;
import io.quartic.common.serdes.ObjectMappers;

import java.io.IOException;
import java.io.InputStream;
import java.util.Iterator;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/**
 * GeoJsonParser is a very forgiving parser of geojson files containing a simple FeatureCollection
 */
public class GeoJsonParser implements Iterator<Feature> {
    enum State {
        START,
        FEATURES,
       END
    }
    private final JsonParser parser;
    private JsonToken token;
    private State state = State.START;


    public GeoJsonParser(InputStream inputStream) throws IOException {
        JsonFactory jsonFactory = new JsonFactory();
        jsonFactory.setCodec(ObjectMappers.OBJECT_MAPPER);
        parser = jsonFactory.createParser(inputStream);
    }

    private void until(JsonToken expected, String text) throws IOException {
        for (token = parser.nextToken() ; token != null; token = parser.nextToken()) {
            if (token.equals(expected) && parser.getText().equals(text)) {
                return;
            }
        }
        throw new IOException("reached end of input searching for: " + expected + "[text=" + text + "]");
    }

    private void expect(JsonToken expected) throws IOException {
        token = parser.nextToken();
        if (!token.equals(expected)) {
            throw new IOException("expected " + expected + " but found " + token);
        }
    }

    private State update() throws IOException {
        switch (state) {
            case START:
                until(JsonToken.FIELD_NAME, "features");
                expect(JsonToken.START_ARRAY);
                expect(JsonToken.START_OBJECT);
                return State.FEATURES;
            case FEATURES:
                if (token == JsonToken.START_OBJECT) {
                    return state;
                } else if (token == JsonToken.END_ARRAY) {
                    return State.END;
                } else throw new IOException("unexpected token: " + token);
            default:
                return state;
        }
    }

    private Feature getNext() throws IOException {
        switch (state) {
            case FEATURES:
                Feature feature = parser.readValueAs(Feature.class);
                token = parser.nextToken();
                return feature;
            default:
                throw new RuntimeException("getNext can only be called in START state");
        }
    }


    @Override
    public boolean hasNext() {
        try {
            state = update();
            return state == State.FEATURES && token == JsonToken.START_OBJECT;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Feature next() {
        try {
            state = update();
            return getNext();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void validate() {
        while (hasNext()) {
            next();
        }
    }

    public Stream<Feature> features() {
        return StreamSupport.stream(
                Spliterators.spliteratorUnknownSize(this, Spliterator.ORDERED),
                false);
    }

}

File: common-core/src/main/kotlin/io/quartic/common/geojson/GeoJsonParser.kt
package io.quartic.common.geojson

import com.fasterxml.jackson.core.JsonFactory
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonToken
import com.fasterxml.jackson.core.JsonToken.*
import io.quartic.common.geojson.GeoJsonParser.State.*
import io.quartic.common.serdes.OBJECT_MAPPER
import java.io.IOException
import java.io.InputStream

/**
 * GeoJsonParser is a very forgiving parser of GeoJSON files containing a simple FeatureCollection.
 */
class GeoJsonParser constructor(inputStream: InputStream) : Iterator<Feature> {
    internal enum class State {
        START,
        FEATURES,
        END
    }

    private val parser: JsonParser
    private var token: JsonToken? = null
    private var state = START

    init {
        val jsonFactory = JsonFactory()
        jsonFactory.codec = OBJECT_MAPPER
        parser = jsonFactory.createParser(inputStream)
    }

    override fun hasNext(): Boolean {
        state = update()
        return state == FEATURES && token == START_OBJECT
    }

    override fun next(): Feature {
        state = update()
        when (state) {
            FEATURES -> {
                val feature = parser.readValueAs(Feature::class.java)
                token = parser.nextToken()
                return feature
            }
            else -> throw RuntimeException("Internal inconsistency - not parsing FEATURES")
        }
    }

    fun validate() {
        while (hasNext()) {
            next()
        }
    }

    private fun update() = when (state) {
        START -> {
            until(FIELD_NAME, "features")
            expect(START_ARRAY)
            expect(START_OBJECT)
            FEATURES
        }
        FEATURES -> {
            when (token) {
                START_OBJECT -> state
                END_ARRAY -> END
                else -> throw IOException("Unexpected token: " + token!!)
            }
        }
        END -> state
    }

    private fun until(expected: JsonToken, text: String) {
        token = parser.nextToken()
        while (token != null) {
            if (token == expected && parser.text == text) {
                return
            }
            token = parser.nextToken()
        }
        throw IOException("Reached end of input searching for: $expected[text=$text]")
    }

    private fun expect(expected: JsonToken) {
        token = parser.nextToken()
        if (token != expected) {
            throw IOException("Expected $expected but found $token")
        }
    }
}
